r/tests/testthat/test-dplyr-funcs-datetime.R

```diff
@@ -650,6 +658,345 @@ test_that("extract yday from date", {
+      ) %>%
+      collect(),
+    test_df_v2
+  )
+
+  compare_dplyr_binding(
+    .input %>%
+      mutate(
+        out_1 = ceiling_date(datetime, "month", change_on_boundary = FALSE),
+        out_2 = ceiling_date(datetime, "quarter", change_on_boundary = FALSE),
+        out_3 = ceiling_date(datetime, "year", change_on_boundary = FALSE),
+      ) %>%
+      collect(),
+    test_df_v2
+  )
+})
+
+
+test_that("change_on_boundary is respected in ceiling_time", {
+
+  boundary_times <- tibble::tibble(
+    datetime = as.POSIXct(strptime(c(
+      "2022-05-10 00:00:00", # boundary for week (Sunday / week_start = 7)
+      "2022-03-10 00:00:00", # boundary for: day, hour, minute, second, millisecond
+      "2022-03-10 00:00:01", # boundary for: second, millisecond
+      "2022-03-10 00:01:00", # boundary for: second, millisecond, minute
+      "2022-03-10 01:00:00"  # boundary for: second, millisecond, minute, hour
+    ), tz="UTC", format = "%F %T"))
+  )
+
+  compare_dplyr_binding(
+    .input %>%
+      mutate(
+        out_1 = ceiling_date(datetime, "day"),
+        out_2 = ceiling_date(datetime, "hour"),
+        out_3 = ceiling_date(datetime, "minute"),
+        out_4 = ceiling_date(datetime, "second"),
+        out_5 = ceiling_date(datetime, ".001 second")
+      ) %>%
+      collect(),
+    boundary_times
+  )
+
+  compare_dplyr_binding(
+    .input %>%
+      mutate(
+        out_1 = ceiling_date(datetime, "day", change_on_boundary = FALSE),
+        out_2 = ceiling_date(datetime, "hour", change_on_boundary = FALSE),
+        out_3 = ceiling_date(datetime, "minute", change_on_boundary = FALSE),
+        out_4 = ceiling_date(datetime, "second", change_on_boundary = FALSE),
+        out_5 = ceiling_date(datetime, ".001 second", change_on_boundary = FALSE)
+      ) %>%
+      collect(),
+    boundary_times
+  )
+
+  compare_dplyr_binding(
+    .input %>%
+      mutate(
+        out_1 = ceiling_date(datetime, "day", change_on_boundary = TRUE),
+        out_2 = ceiling_date(datetime, "hour", change_on_boundary = TRUE),
+        out_3 = ceiling_date(datetime, "minute", change_on_boundary = TRUE),
+        out_4 = ceiling_date(datetime, "second", change_on_boundary = TRUE),
+        out_5 = ceiling_date(datetime, ".001 second", change_on_boundary = TRUE)
+      ) %>%
+      collect(),
+    boundary_times
+  )
+
+  boundary_dates <- tibble::tibble(
+    date = as.Date(c(
+      "2001-05-10", # regular day
+      "2002-05-10", # regular day
+      "2003-05-10", # regular day
+      "2004-05-10", # regular day
+      "2005-05-10", # regular day
+      "2006-05-10", # regular day
+      "2007-05-10"  # regular day
+    ))
+  )
+
+  compare_dplyr_binding(
+    .input %>%
+      mutate(
+        out_1 = ceiling_date(date, "day", change_on_boundary = FALSE),
+        out_2 = ceiling_date(date, "day", change_on_boundary = TRUE),
+        out_3 = ceiling_date(date, "day")
+      ) %>%
+      collect(),
+    boundary_dates
+  )
```

Comment:
Hi @rok. Making progress with this, but pausing to ping you because this test currently throws a very strange error. The input here is a vector of dates with years 2001, 2002, 2003, etc. The output should contain the same vectors, because the rounding unit here is just to the nearest day, i.e.,

```
2001-05-10,
2002-05-10,
2003-05-10,
2004-05-10,
etc
```

what we actually get is the correct answer but with the unix epoch interleaved, i.e.,

```
2001-05-10,
1970-01-01,
2002-05-10,
1970-01-01,
2003-05-10,
etc
```

My guess is that this must be something to do with how the data buffer is being parsed? I'm guessing a single 64-bit date in the C++ library has been treated as if it is two 32-bit dates when it's being pulled back into R. Any thoughts on what might be going on here?

Reply:
Your reasoning with date64 accidentally becoming two date32s makes sense, but I've no idea how it might be happening. I'll take a look if I can reproduce this in C++ tests.

Reply:
Yeah I see the same in C++. With the following change:
```
+++ b/cpp/src/arrow/compute/kernels/scalar_temporal_test.cc
@@ -2894,6 +2894,16 @@ TEST_F(ScalarTemporalTest, TestRoundTemporal) {
   CheckScalarUnary(op, unit, times, unit, round_15_years, &round_to_15_years);
 }
 
+TEST_F(ScalarTemporalTest, TestCeilFloorRoundTemporalDate) {
+  RoundTemporalOptions round_to_2_hours = RoundTemporalOptions(2, CalendarUnit::HOUR);
+  const char* date32s = R"([0, 11016, -25932, null])";
+  const char* date64s = R"([0, 951782400000, -2240524800000, null])";
+  auto dates32 = ArrayFromJSON(date32(), date32s);
+  auto dates64 = ArrayFromJSON(date64(), date64s);
+  CheckScalarUnary("ceil_temporal", dates64, dates64, &round_to_2_hours);
+  CheckScalarUnary("ceil_temporal", dates32, dates32, &round_to_2_hours);
+}
```

I get:
```
Got:
  [
    [
      1970-01-01,
      1970-01-01,
      2000-02-29,
      null
    ]
  ]
Expected:
  [
    [
      1970-01-01
    ],
    [
      2000-02-29,
      1899-01-01,
      null
    ]
  ]
```

I'll open a Jira for this and I suggest you disable `date32` until it is resolved.

Reply:
Okay, will do! Thanks 🙂

Reply:
This seems to happen for `time32` too. Opened [ARROW-16142](https://issues.apache.org/jira/browse/ARROW-16142).